docs/design-documents/20221027-builtin-config-validation.md

```diff
@@ -0,0 +1,144 @@
+      string value = 2;
+    }
+
+    // Type shows the parameter type.
+    enum Type {
+      TYPE_UNSPECIFIED = 0;
+      // Parameter is a string.
+      TYPE_STRING = 1;
+      // Parameter is a number.
+      TYPE_NUMBER = 2;
+      // Parameter is a boolean.
+      TYPE_BOOL = 3;
+      // Parameter is a file.
+      TYPE_FILE = 4;
+      // Parameter is a duration.
+      TYPE_DURATION = 5;
+    }
+
+    string description = 1;
+    string default = 2;
+    Type type = 3;
+    repeated Validation validations = 4;
+  }
+
+  string name = 1;
+  string summary = 2;
+  string description = 3;
+  string version = 4;
+  string author = 5;
+  map<string, Parameter> destination_params = 6;
+  map<string, Parameter> source_params = 7;
+}
+```
+
+For now, this is only used when listing the plugins in Conduit. All these validation options are not yet implemented and
+are not exposed by the SDK.
+
+## Scope
+
+Implementing the validation options provided by the proto file. So, giving the developer the option to specify
+validations for each parameter and Conduit will make sure to run the validations.
+
+Validating the type of the parameter, we have 5 types supported in the proto design {string, number, bool, file, and duration}
+
+Providing a utility function to generate the config map for the `Parameters` function from a config struct. This is not 
+mandatory for the scope of this feature, but it would be a nice to have and would make the developing experience for connectors easier.
+
+## Questions:
+
+**Q**: Should we implement the validations from the SDK or Conduit side?
+
+**A**: If we implement it from Conduit side, this means we can add more validations in the future without changing the SDK.
+However, implementing it from the SDK side means that all the validations will happen on the connector side, using the 
+Config function, and the developer will still have the ability to add custom validations for some specific cases that 
+the SDK wouldn't cover. So we decided on implementing validating from the SDK side.
+
+**Q**: Should generating the Configurations from a Go struct be in the scope of this feature?
+
+**A**: We decided to add the generation as a subtask for now, and see how much time left we have for it. We agreed that this
+would be super helpful for developers and easier to use, and will continue adding to the feature in the future.
+
+## Implementation
+
+Implementing this feature consists of three main steps:
+
+1. Give developers the ability to specify validations needed for each parameter. (manually)
+   Params should look something like:
+
+```go
+SourceParams: []sdk.Parameter{
+    {
+      Name: "param",
+      Type: sdk.ParameterTypeNumber,
+      Validations: []sdk.Validation{
+        sdk.Validation{
+          Type: sdk.ValidationRequired,
+        },
+        sdk.Validation{
+          Type: sdk.ValidationGreaterThan,
+          Value: 0,
+        },
+      }
```

Comment:
Yes this looks much neater! we're still going to have to convert this to the first format, because that's the one needed to match the proto message design, which will happen on the SDK and the developer won't have to worry about it.

Reply:
I think I'm missing something... I read in this doc that the validations will be done in the SDK. And the proto files mean we're going to send them to Conduit. Why though?

Reply:
We're sending them over so the UI can do the validations and provide a better UX.

Reply:
So if a configuration has a combination of built-in and custom validations, the UI would be doing the built-in validations, and then the custom validations will be done in the connector itself?

Reply:
yes! If we see that this shouldn't be the case, then I'm thinking of one simple solution for it, the UI could also use the `validate` endpoint for Conduit, since it calls the `Configure` method from the connector.
But in general, I'm on the fence about whether we should validate both the builtin and the custom validations or not. WDYT?

Reply:
Some thoughts on hitting `validate` after each config value is entered:
- In case of a standalone plugin, the `validate` endpoint will start and stop plugin each time, it's probably not the fastest endpoint (might not be an issue, but we should check how long it takes on average, just to be sure).
- The returned errors need to allow the UI to figure out which field is still wrong (we don't want to create generic pop-ups for these validations, right?).
  - We would also need to be smart about which errors to ignore, otherwise all fields would turn red once the user switches away from the first field.
- The `validate` endpoint will execute the standard validations as well. Do we even need to tell the UI about the standard validations at that point?

Reply:
discussed during conduit sync 22/11/2022, the UI will validate the builtlin validations for each configuration while the user is typing (validations are found in the same struct that has the list of the fields). When the user submits the form, the UI will try and create the connector (which will validate both the builtin and the custom validation), and will show an error for the user if it occurs. @jmar910